Mathlib/Algebra/Group/Opposite.lean

```diff
@@ -0,0 +1,597 @@
+    intCast := fun n => op n,
+    intCast_ofNat := fun n => show op ((n : ℤ) : α) = op (n : α) by rw [Int.cast_ofNat],
+    intCast_negSucc := fun n =>
+      show op _ = op (-unop (op ((n + 1 : ℕ) : α))) by erw [unop_op, Int.cast_negSucc] }
+
+instance [AddCommGroup α] : AddCommGroup αᵐᵒᵖ :=
+  unop_injective.addCommGroup _ rfl (fun _ _ => rfl) (fun _ => rfl) (fun _ _ => rfl)
+    (fun _ _ => rfl) fun _ _ => rfl
+
+/-!
+### Multiplicative structures on `αᵐᵒᵖ`
+
+We also generate additive structures on `αᵃᵒᵖ` using `to_additive`
+-/
+
+
+@[to_additive]
+instance [Semigroup α] : Semigroup αᵐᵒᵖ :=
+  { instMulMulOpposite α with
+    mul_assoc := fun x y z => unop_injective <| Eq.symm <| mul_assoc (unop z) (unop y) (unop x) }
+
+@[to_additive]
+instance [RightCancelSemigroup α] : LeftCancelSemigroup αᵐᵒᵖ :=
+  { instSemigroupMulOpposite α with
+    mul_left_cancel := fun _ _ _ H => unop_injective <| mul_right_cancel <| op_injective H }
+
+@[to_additive]
+instance [LeftCancelSemigroup α] : RightCancelSemigroup αᵐᵒᵖ :=
+  { instSemigroupMulOpposite α with
+    mul_right_cancel := fun _ _ _ H => unop_injective <| mul_left_cancel <| op_injective H }
+
+@[to_additive]
+instance [CommSemigroup α] : CommSemigroup αᵐᵒᵖ :=
+  { instSemigroupMulOpposite α with
+    mul_comm := fun x y => unop_injective <| mul_comm (unop y) (unop x) }
+
+@[to_additive]
+instance [MulOneClass α] : MulOneClass αᵐᵒᵖ :=
+  { instMulMulOpposite α, instOneMulOpposite α with
+    one_mul := fun x => unop_injective <| mul_one <| unop x,
+    mul_one := fun x => unop_injective <| one_mul <| unop x }
+
+@[to_additive]
+instance [Monoid α] : Monoid αᵐᵒᵖ :=
+  { instSemigroupMulOpposite α, instMulOneClassMulOpposite α with
+    npow := fun n x => op <| x.unop ^ n,
+    npow_zero := fun x => unop_injective <| Monoid.npow_zero x.unop,
+    npow_succ := fun n x => unop_injective <| pow_succ' x.unop n }
+
+@[to_additive]
+instance [RightCancelMonoid α] : LeftCancelMonoid αᵐᵒᵖ :=
+  { instLeftCancelSemigroupMulOpposite α, instMonoidMulOpposite α with }
+
+@[to_additive]
+instance [LeftCancelMonoid α] : RightCancelMonoid αᵐᵒᵖ :=
+  { instRightCancelSemigroupMulOpposite α, instMonoidMulOpposite α with }
+
+@[to_additive]
+instance [CancelMonoid α] : CancelMonoid αᵐᵒᵖ :=
+  { instRightCancelMonoidMulOpposite α, instLeftCancelMonoidMulOpposite α with }
+
+@[to_additive]
+instance [CommMonoid α] : CommMonoid αᵐᵒᵖ :=
+  { instMonoidMulOpposite α, instCommSemigroupMulOpposite α with }
+
+@[to_additive]
+instance [CancelCommMonoid α] : CancelCommMonoid αᵐᵒᵖ :=
+  { instCancelMonoidMulOpposite α, instCommMonoidMulOpposite α with }
+
+-- Porting note: We have to add the `div` definition because of an error occuring in `to_additive`
+-- Declarations if the definition isn't included.
+@[to_additive AddOpposite.subNegMonoid]
+instance [DivInvMonoid α] : DivInvMonoid αᵐᵒᵖ :=
+  { instMonoidMulOpposite α, instInvMulOpposite α with
+    div := fun x y => x * y⁻¹
```

Comment:
after #952 is merged you can remove this again

Reply:
Done!